File: backend-node/src/research/edge-metadata.ts
import type { Edge, EdgeMetadata, EdgeSource, Graph, Support } from './schema';

type EdgeCore = Omit<Edge, keyof EdgeMetadata>;
type LedgerData = Pick<Graph, 'claims' | 'sources' | 'created_at' | 'mode'>;
const BASE: Record<Support, number> = { directly_supported: 0.70, strongly_inferred: 0.50, weakly_inferred: 0.30, disputed: 0.70, user_asserted: 0.25, unresolved: 0 };
const MAX_SCORE = 0.95;
const DISPUTED_CAP = 0.25;
const YEAR_MS = 365.25 * 24 * 60 * 60 * 1000;
function latestDate(values: (string | null | undefined)[]): string | null {
  const times = values.filter((v): v is string => !!v).map(v => Date.parse(v)).filter(Number.isFinite);
  return times.length ? new Date(Math.max(...times)).toISOString() : null;
}

/** Pure evidence-derived projection; never asks the model for a confidence percentage.
 * Time is pinned to recorded observations, so reloading an old graph cannot age its score.
 */
export function edgeMetadata(edge: EdgeCore, graph: LedgerData): EdgeMetadata {
  const claimIds = new Set(edge.claim_ids);
  const claims = graph.claims.filter(c => claimIds.has(c.id));
  const evidence = claims.flatMap(c => c.evidence);
  const sourceIds = [...new Set(evidence.map(e => e.source_id))].sort();
  const source: EdgeSource[] = sourceIds.flatMap(sid => {
    const found = graph.sources.find(s => s.id === sid);
    return found ? [{ ...found, support_types: [...new Set(evidence.filter(e => e.source_id === sid).map(e => e.support_type))].sort() }] : [];
  });
  const missing_claim_ids = edge.claim_ids.filter(cid => !claims.some(c => c.id === cid));
  const missing_source_ids = sourceIds.filter(sid => !source.some(s => s.id === sid));
  const observed = latestDate(claims.flatMap(c => [c.observed_at, ...c.evidence.map(e => e.extracted_at)]))
    ?? latestDate(source.map(s => s.retrieved_at)) ?? latestDate([graph.created_at]);
  const positives = source.filter(s => s.support_types.includes('supports'));
  const web = positives.filter(s => s.kind !== 'upload' && evidence.some(e => e.source_id === s.id && e.support_type === 'supports' && !!e.span.trim()));
  // Connected components of shared family OR shared document hash: copied pages
  // across different hosts must not earn extra corroboration, even transitively.
  const groups: { families: Set<string>; hashes: Set<string> }[] = [];
  for (const s of web) {
    const family = s.source_family_id || 'unknown_family';
    const content = s.content_hash || 'unknown_content';
    const matches = groups.filter(g => g.families.has(family) || g.hashes.has(content));
    const merged = { families: new Set([family]), hashes: new Set([content]) };
    for (const g of matches) { g.families.forEach(f => merged.families.add(f)); g.hashes.forEach(h => merged.hashes.add(h)); groups.splice(groups.indexOf(g), 1); }
    groups.push(merged);
  }
  const notes = ['Heuristic evidence score, not a calibrated probability or a measure of supply-chain completeness.', 'Family independence uses stored source-family IDs and content hashes; edited syndication may remain undetected.'];
  const hasSupport = evidence.some(e => e.support_type === 'supports' && !!e.span.trim() && positives.some(s => s.id === e.source_id));
  const disputed = edge.support_label === 'disputed' || evidence.some(e => e.support_type === 'contradicts');
  let base = hasSupport ? BASE[edge.support_label] : 0;
  if (!web.length) base = Math.min(base, BASE.user_asserted);
  const corroboration = Math.min(2, Math.max(0, groups.length - 1)) * 0.10;
  let freshness = 0;
  if (web.length) {
    const published = latestDate(web.map(s => s.published_at).filter(d => d && observed && Date.parse(d) <= Date.parse(observed)));
    if (!published || !observed) { freshness = -0.05; notes.push('Supporting publication date unknown or in the future; retrieval time is not publication time.'); }
    else {
      const age = (Date.parse(observed) - Date.parse(published)) / YEAR_MS;
      freshness = age > 5 ? -0.20 : age > 2 ? -0.10 : 0;
      if (freshness) notes.push('Older supporting publication reduces freshness; historical evidence may still be correct.');
    }
  } else notes.push('User assertion or missing web support; not independently corroborated.');
  const contradiction = disputed ? -0.45 : 0;
  if (disputed) notes.push('Contrary evidence caps confidence at 0.25; this does not estimate a probability that either side is true.');
  if (graph.mode === 'replay') notes.push('Replay uses curated verdicts; the score does not imply a live model review.');
  const incomplete = !hasSupport || missing_claim_ids.length > 0 || missing_source_ids.length > 0;
  if (incomplete) notes.push('Missing evidence lineage forces the score to zero.');
  const confidence = incomplete ? 0 : Math.round(Math.max(0, Math.min(disputed ? DISPUTED_CAP : MAX_SCORE, base + corroboration + freshness + contradiction)) * 1000) / 1000;
  return {
    source, date: observed?.slice(0, 10) ?? null, time: observed?.slice(11) ?? null, confidence,
    confidence_details: {
      method: { name: 'edge_evidence_v1', params: { scope: 'evidence_for_stated_relation', scale: '0-1', max_score: MAX_SCORE, disputed_cap: DISPUTED_CAP } },
      factors: { base, corroboration, freshness, contradiction },
      data_quality: { calibrated: false, replay: graph.mode === 'replay', supporting_families: groups.length, missing_claim_ids, missing_source_ids, notes },
      evaluated_at: observed,
    },
  };
}
export function annotateEdge<T extends EdgeCore>(edge: T, graph: LedgerData): T & EdgeMetadata { return Object.assign(edge, edgeMetadata(edge, graph)); }
export function hasEdgeMetadata(edge: EdgeCore): edge is Edge {
  const value = edge as Partial<Edge>;
  return Array.isArray(value.source) && value.date !== undefined && value.time !== undefined && typeof value.confidence === 'number' && !!value.confidence_details;
}
/** Upgrade pre-metadata files on read without rewriting their original audit records. */
export function ensureGraphEdgeMetadata(graph: Graph): Graph {
  for (const edge of graph.edges) if (!hasEdgeMetadata(edge)) annotateEdge(edge, graph);
  return graph;
}

File: backend/app/edge_metadata.py
"""Edge provenance and a deterministic evidence score (edge_evidence_v1).

Pure projection of the claim ledger; never asks a model for a confidence percentage. The
scoring clock is pinned to recorded evidence observations, so reloading an old graph cannot
age its score. Ported from the Node backend's edge-metadata module; see EDGE_METADATA.md.
"""

import math
from datetime import UTC, datetime

BASE = {
    "directly_supported": 0.70,
    "strongly_inferred": 0.50,
    "weakly_inferred": 0.30,
    "disputed": 0.70,
    "user_asserted": 0.25,
    "unresolved": 0.0,
}
MAX_SCORE = 0.95
DISPUTED_CAP = 0.25
YEAR_SECONDS = 365.25 * 24 * 60 * 60
METHOD = {
    "name": "edge_evidence_v1",
    "params": {
        "scope": "evidence_for_stated_relation",
        "scale": "0-1",
        "max_score": MAX_SCORE,
        "disputed_cap": DISPUTED_CAP,
    },
}


def parse_time(value):
    if not isinstance(value, str) or not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=UTC)
    return parsed.astimezone(UTC)


def latest(values):
    times = [t for t in (parse_time(v) for v in values) if t is not None]
    return max(times) if times else None


def stamp(moment):
    """UTC ISO timestamp with millisecond precision, matching the contract's time pattern."""
    return moment.strftime("%Y-%m-%dT%H:%M:%S.") + f"{moment.microsecond // 1000:03d}Z"


def edge_metadata(edge, graph):
    claims_by_id = graph.get("_claims", {})
    claims = [claims_by_id[cid] for cid in edge["claim_ids"] if cid in claims_by_id]
    # Rejected claims are no longer evidence for the relation, though they stay in the ledger.
    active = [c for c in claims if c.get("status") != "rejected"]
    evidence = [e for c in active for e in c.get("evidence", [])]
    inline = {}
    for item in evidence:
        source = item.get("source")
        if isinstance(source, dict) and source.get("id"):
            inline.setdefault(source["id"], source)
    source_ids = sorted({e["source_id"] for e in evidence})
    sources = []
    for sid in source_ids:
        found = inline.get(sid)
        if found is None:
            continue
        types = sorted({e["support_type"] for e in evidence if e["source_id"] == sid})
        sources.append(
            {**{k: v for k, v in found.items() if not k.startswith("_")}, "support_types": types}
        )
    missing_claim_ids = [cid for cid in edge["claim_ids"] if cid not in claims_by_id]
    missing_source_ids = [sid for sid in source_ids if sid not in inline]
    observed = (
        latest([c.get("observed_at") for c in active] + [e.get("extracted_at") for e in evidence])
        or latest([s.get("retrieved_at") for s in sources])
        or latest([graph.get("created_at")])
    )
    positives = [s for s in sources if "supports" in s["support_types"]]
    web = [
        s
        for s in positives
        if s.get("kind") != "upload"
        and any(
            e["source_id"] == s["id"]
            and e["support_type"] == "supports"
            and e.get("span", "").strip()
            for e in evidence
        )
    ]
    # Connected components of shared family OR shared document hash: copied pages across
    # different hosts must not earn extra corroboration, even transitively.
    groups = []
    for s in web:
        family = s.get("source_family_id") or "unknown_family"
        content = s.get("content_hash") or "unknown_content"
        merged = {"families": {family}, "hashes": {content}}
        remaining = []
        for group in groups:
            if family in group["families"] or content in group["hashes"]:
                merged["families"] |= group["families"]
                merged["hashes"] |= group["hashes"]
            else:
                remaining.append(group)
        groups = remaining + [merged]
    notes = [
        "Heuristic evidence score, not a calibrated probability or a measure of supply-chain completeness.",
        "Family independence uses stored source-family IDs and content hashes; edited syndication may remain undetected.",
    ]
    positive_ids = {s["id"] for s in positives}
    has_support = any(
        e["support_type"] == "supports"
        and e.get("span", "").strip()
        and e["source_id"] in positive_ids
        for e in evidence
    )
    disputed = (
        edge.get("support_label") == "disputed"
        or any(e["support_type"] == "contradicts" for e in evidence)
        or any(
            c.get("status") == "disputed" or c.get("support_label") == "disputed" for c in active
        )
    )
    base = BASE.get(edge.get("support_label"), 0.0) if has_support else 0.0
    if not web:
        base = min(base, BASE["user_asserted"])
    corroboration = min(2, max(0, len(groups) - 1)) * 0.10
    freshness = 0.0
    if web:
        published = latest(
            [
                s.get("published_at")
                for s in web
                if s.get("published_at")
                and observed
                and parse_time(s["published_at"]) is not None
                and parse_time(s["published_at"]) <= observed
            ]
        )
        if published is None or observed is None:
            freshness = -0.05
            notes.append(
                "Supporting publication date unknown or in the future; retrieval time is not publication time."
            )
        else:
            age = (observed - published).total_seconds() / YEAR_SECONDS
            freshness = -0.20 if age > 5 else -0.10 if age > 2 else 0.0
            if freshness:
                notes.append(
                    "Older supporting publication reduces freshness; historical evidence may still be correct."
                )
    else:
        notes.append("User assertion or missing web support; not independently corroborated.")
    contradiction = -0.45 if disputed else 0.0
    if disputed:
        notes.append(
            "Contrary evidence caps confidence at 0.25; this does not estimate a probability that either side is true."
        )
    if graph.get("mode") == "replay":
        notes.append("Replay uses curated verdicts; the score does not imply a live model review.")
    incomplete = not has_support or missing_claim_ids or missing_source_ids
    if incomplete:
        notes.append("Missing evidence lineage forces the score to zero.")
    if incomplete:
        confidence = 0.0
    else:
        total = base + corroboration + freshness + contradiction
        confidence = max(0.0, min(DISPUTED_CAP if disputed else MAX_SCORE, total))
        confidence = math.floor(confidence * 1000 + 0.5) / 1000
    evaluated = stamp(observed) if observed else None
    return {
        "source": sources,
        "date": evaluated[:10] if evaluated else None,
        "time": evaluated[11:] if evaluated else None,
        "confidence": confidence,
        "confidence_details": {
            "method": METHOD,
            "factors": {
                "base": base,
                "corroboration": corroboration,
                "freshness": freshness,
                "contradiction": contradiction,
            },
            "data_quality": {
                "calibrated": False,
                "replay": graph.get("mode") == "replay",
                "supporting_families": len(groups),
                "missing_claim_ids": missing_claim_ids,
                "missing_source_ids": missing_source_ids,
                "notes": notes,
            },
            "evaluated_at": evaluated,
        },
    }


def annotate_edge(edge, graph):
    edge.update(edge_metadata(edge, graph))
    return edge


def has_edge_metadata(edge):
    return (
        isinstance(edge.get("source"), list)
        and "date" in edge
        and "time" in edge
        and isinstance(edge.get("confidence"), int | float)
        and bool(edge.get("confidence_details"))
    )


def ensure_edge_metadata(graph):
    """Upgrade snapshots saved before this feature on read, without rewriting them."""
    for edge in graph.get("edges", []):
        if not has_edge_metadata(edge):
            annotate_edge(edge, graph)
    return graph
